### ISSUE TITLE
training GAE, loss is fixed with 34.5388

### ISSUE THREAD

**Issue by BillChow117** [rusty1s/pytorch_geometric#1165]
## ❓ Questions & Help

Hi! I am trying to train a GAE to obtain embeddings of a graph.
Here is my code. The "data" only includes feature matrix "x" and  edge index "edge_index". The loss is always 34.5388. I am so confused.   

```python
import torch
import torch.nn as nn
import torchvision
import torch.nn.functional as F
from torch_geometric.nn import GAE, VGAE, GCNConv, GraphConv
from torch_geometric.data import Data
from tqdm import trange

class Encoder1(nn.Module):
    def __init__(self, in_channels = None, out_channels = None):
        super(Encoder1, self).__init__()
        self.conv1 = GCNConv(in_channels,2 * out_channels, cached = True)
        self.conv2 = GCNConv(2 * out_channels, out_channels, cached = True)
    
    def forward(self, x, edge_index):        
        
        z = F.relu(self.conv1(x, edge_index))
        z = self.conv2(z, edge_index)   
        
        return z

class Encoder2(nn.Module):
    def __init__(self, in_channels = None, out_channels = None):
        super(Encoder2, self).__init__()
        self.conv1 = GraphConv(in_channels,2 * out_channels)
        self.conv2 = GraphConv(2 * out_channels, out_channels)
    
    def forward(self, x, edge_index):        
        
        z = F.relu(self.conv1(x, edge_index))
        z = self.conv2(z, edge_index)   
        return z


def get_model(lr = 0.001, GCN = 1, in_channels = None, Z_num = 16, device = 'cuda'):
    if GCN == 1:
        encoder = Encoder1(in_channels = in_channels, out_channels = Z_num)
    elif GCN == 2:
        encoder = Encoder2(in_channels = in_channels, out_channels = Z_num)

    model = GAE(encoder = encoder).to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr = lr)

    return model, optimizer

def train(model, optimizer, data):
    model.train()
    optimizer.zero_grad()
    z = model.encode(data.x, data.edge_index)
    loss = model.recon_loss(z, data.edge_index)
    #loss = recon_loss + (1 / data.num_nodes) * model.kl_loss()
    loss.backward()
    optimizer.step()
    return loss

def get_Z(model, optimizer, epoch, data):
    model.reset_parameters()
    for i in trange(epoch):
        loss = train(model, optimizer, data)

    model.eval()
    z = model.encode(data.x, data.edge_index)
    return z, loss


device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
datapath = 'test/result/S_8/imagedata/000001.pt'
data = torch.load(datapath).to(device)
inchannels = data.num_features
model, optimizer = get_model(in_channels = inchannels, device = device)
Z, loss = get_Z(model, optimizer, 200, data)
print(Z, loss)
```

**Comment by rusty1s** [comment]
The code looks correct to me, so I assume that may be data-dependent bug, *e.g.* NaNs occuring in the backward pass. You can ensure that by using `torch.detect_anomaly()`.

**Comment by BillChow117** [comment]
> The code looks correct to me, so I assume that may be data-dependent bug, _e.g._ NaNs occuring in the backward pass. You can ensure that this is not the case by using `torch.detect_anomaly()`.

I'm sorry it took so long to reply. It seems to be caused by my data. There is a small chance that the loss will drop to slightly more than 1, but in most cases it is 34.5388. Is it possible that it was caused by a lot of 0s and -1s in my feature matrix？

**Comment by BillChow117** [comment]
I could only find ```torch.autograde.detect_anomaly()```in pytorch doc and followed it to use the function. But the output remained unchanged. I also tried to remove those -1s in the data. But it doesn't help.
I changed these two functions:
```python
def train(model, optimizer, data):
    model.train()
    optimizer.zero_grad()
    with torch.autograd.detect_anomaly():
        z = model.encode(data.x, data.edge_index)
        loss = model.recon_loss(z, data.edge_index)
        #loss = recon_loss + (1 / data.num_nodes) * model.kl_loss()
        loss.backward()
        optimizer.step()
    return loss

def get_Z(model, optimizer, epoch, data):
    model.reset_parameters()
    torch.autograd.set_detect_anomaly(True)
    for i in trange(epoch):
        loss = train(model, optimizer, data)

    model.eval()
    z = model.encode(data.x, data.edge_index)
    return z, loss
```
and I got output which is same as it was:
tensor([[ -3.6404,   7.5190,  -4.0397,  ...,  -5.0865,   0.9669,   7.2040],
        [ -4.3972,   8.6370,  -5.1425,  ...,  -5.9306,   1.4153,   8.3739],
        [ -4.3795,   7.7792,  -5.6073,  ...,  -5.4598,   1.9261,   7.7511],
        ...,
        [-31.9277,  42.0693, -65.0303,  ...,  -2.4038, -10.6528,  47.9778],
        [-33.3238,  43.2352, -67.9586,  ...,  -1.8205, -10.8781,  49.8694],
        [-28.1205,  36.1329, -57.3122,  ...,  -1.2419,  -9.0839,  41.9853]],
       device='cuda:0', grad_fn=<AddBackward0>) tensor(34.5388, device='cuda:0', grad_fn=<AddBackward0>)

**Comment by zequan-wu** [comment]
@BillChow117   Hi! I also encountered this problem when I was training the GAE model. The loss was always 34.583, and then dropped to 1.38** instantly. Have you solved this problem now? My training code is basically the same as yours. Is it a code problem?

**Comment by wenyuhaokikika** [comment]
I also encountered this problem.
Each dataset is different, so the required parameters are also different
please see sigmoid(z@z.t()),it may all eq 1.
i fix it by reduce LR=0.005 instead of 0.01 or higher.

**Comment by hadirkhan10** [comment]
I was also facing a similar problem. The sigmoid was setting all the probabilities to 1 and the model wasn't learning anything. I resolved it by normalizing my features. Having features in a larger scale my overpower the output produced by the model and the sigmoid just sets everything to 1
